src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1954,200 @@ MutableBigInteger sqrt() {
         }
     }
 
+    /**
+     * Calculate the integer square root {@code floor(sqrt(this))} and the remainder
+     * if needed, where {@code sqrt(.)} denotes the mathematical square root.
+     * The contents of {@code this} are <em>not</em> changed.
+     * The value of {@code this} is assumed to be non-negative.
+     *
+     * @return the integer square root of {@code this} and the remainder if needed
+     * @throws ArithmeticException if the value returned by {@code bitLength()}
+     * overflows the range of {@code int}.
+     */
+    MutableBigInteger[] sqrtRem(boolean needRemainder) {
+        // Special cases.
+        if (this.intLen <= 2) {
+            final long x = this.toLong(); // unsigned
+
+            /* For every long value s in [0, 2^32) such that x == s * s,
+             * it is true that s == (long) Math.sqrt(x >= 0 ? x : x + 0x1p64).
+             * This means that Math.sqrt() returns the correct value for every perfect square,
+             * so the value returned by Math.sqrt() for a long value in the range [0, 2^64)
+             * is either correct, or rounded up by one if the value is too high
+             * and too close to the next perfect square.
+             */
```

Comment:
Contrary to my previous believe and own experiments, I now think this code is incorrect.

Let `long t = 3037000503L` and `long x = t * t`. The code computes `long s == 3037000502L`, an underestimate of the correct square root `t` by 1. Underestimates are neither detected nor corrected.
Of course, the corresponding remainder `long r = x - s * s`, namely `r = 6074001005L`, is just barely too large as it does _not_ meet `r <= 2 * s`.

Reply:
In fact, if you run this code:
`long limit = 1L << 32;
 for (long n = 0; n < limit; n++) {
      long x = n * n;
      if (n != (long) Math.sqrt(x >= 0 ? x : x + 0x1p64)) {
          System.out.println(n);
      }
}`

now you find a lot of counterexamples. The question is: why, until recently, if I run the same code I could not find a counterexample?

Reply:
I hope these errors are not due to an implementation change in the virtual machine instructions...

Reply:
There are no counterexamples for perfect squares if you write `long s = (long) Math.rint(Math.sqrt(x >= 0 ? x : x + 0x1p64));`.

Reply:
@rgiulietti Is it normal that the same code did not find counterexamples until recently, and now it finds them?

Reply:
I tried on older release, they all agree.

Reply:
@rgiulietti  This is so strange... anyway, I tried also `long x = n * n`, `long s = (long) Math.ceil(Math.sqrt(x >= 0 ? x : x + 0x1p64))` with the test `s < n`, which I think it's more mathematically natural, and also this never fails for perfect squares.

Reply:
Also, this avoids a test
```
            if (Long.compareUnsigned(x, s * s - 1) <= 0) {  // benign over- and underflows
                s--;
            }
```

Reply:
Sorry, disregard the above as it doesn't work for x = 0.

Reply:
Can try with the old release and the incorrect code again?
If the results disagree with newer releases then I'd be interested in which release you were using, as to analyze the generated code and possibly file a bug report for the JIT compiler team.

Reply:
@rgiulietti If I remember correctly, I had tested the code in a bit of a hurry on an online compiler, that of W3Schools, so probably I could have made a mistake writing the test code.

Reply:
@rgiulietti The strangest thing is that we were both convinced that we had correctly tested the code on all values, without finding counterexamples...

Reply:
I was trying with my own code, the one with the additional `Math.rint()` invocation, and didn't notice that the code in the PR was using a `(long)` conversion directly.
So my throw-away mini-program was slightly different from yours.